### PR TITLE
Bug Fix: resolving issue 268

### PR DESCRIPTION
**Pull request by bdorney**
## Description

Resolves issue #268 

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

Breaking because return type of `HwAMC:: getTriggerLinkStatus(...)` has changed.

## Motivation and Context

Issue #268

## How Has This Been Tested?

https://github.com/cms-gem-daq-project/vfatqc-python-scripts/pull/241

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by mexanick — DISMISSED** [review]

File: gempython/tools/amc_user_functions_xhal.py [hunk 1]
```diff
@@ -431,55 +433,54 @@ def getTriggerLinkStatus(self,printSummary=False, checkCSCTrigLink=False, ohMask
...
-                    print("LINK1_MISSED_COMMA_CNT		{0}{1}{2}".format(colors.RED if (linkResult[index+1] > 0) else colors.GREEN,linkResult[index+1],colors.ENDC))
-                    print("LINK0_OVERFLOW_CNT		{0}{1}{2}".format(colors.RED if (linkResult[index+2] > 0) else colors.GREEN,linkResult[index+2],colors.ENDC))
-                    print("LINK1_OVERFLOW_CNT		{0}{1}{2}".format(colors.RED if (linkResult[index+3] > 0) else colors.GREEN,linkResult[index+3],colors.ENDC))
-                    print("LINK0_UNDERFLOW_CNT		{0}{1}{2}".format(colors.RED if (linkResult[index+4] > 0) else colors.GREEN,linkResult[index+4],colors.ENDC))
-                    print("LINK1_UNDERFLOW_CNT		{0}{1}{2}".format(colors.RED if (linkResult[index+5] > 0) else colors.GREEN,linkResult[index+5],colors.ENDC))
-                    print("LINK0_SBIT_OVERFLOW_CNT		{0}{1}{2}".format(colors.RED if (linkResult[index+6] > 0) else colors.GREEN,linkResult[index+6],colors.ENDC))
-                    print("LINK1_SBIT_OVERFLOW_CNT		{0}{1}{2}".format(colors.RED if (linkResult[index+7] > 0) else colors.GREEN,linkResult[index+7],colors.ENDC))
+                    print("----------OH{0}----------".format(ohN))
```
Comment (BenjaminRS) [inline gempython/tools/amc_user_functions_xhal.py]: Here I think you mean ohN+1

**Review by BenjaminRS — COMMENTED** [review]
To me this looks fine, except a print out of ohN which should be changed if checking the CSC trigger

File: gempython/tools/amc_user_functions_xhal.py (hunk 1 above)
Comment (bdorney) [reply gempython/tools/amc_user_functions_xhal.py]: Yes I do thanks for catching this.

**Comment by bdorney** [comment]
> To me this looks fine, except a print out of ohN which should be changed if checking the CSC trigger

Issue has been resolved.

**Comment by bdorney** [comment]
Don't merge this, I found another [bug](https://github.com/cms-gem-daq-project/cmsgemos/issues/268#issuecomment-479089059)

**Comment by bdorney** [comment]
> Don't merge this, I found another [bug](https://github.com/cms-gem-daq-project/cmsgemos/issues/268#issuecomment-479089059)

commit `ef93306` has resolved this bug.  I've retested without `checkCSCTrigLink`:

```bash
% testConnectivity.py -d --shelf=1 -s4 -o 0x4c --nPhaseScans=100 --skipDACScan --skipScurve -a
...
...
====================
Step 3: Programming FPGA & Checking FPGA Communication
====================
Initial value to write: 1, register GEM_AMC.TTC.GENERATOR.ENABLE
Initial value to write: 1, register GEM_AMC.TTC.GENERATOR.SINGLE_HARD_RESET
Initial value to write: 0, register GEM_AMC.TTC.GENERATOR.ENABLE
Checking trigger link status:
--=======================================--
-> GEM SYSTEM TRIGGER LINK INFORMATION
--=======================================--

----------OH2----------
LINK0_MISSED_COMMA_CNT          0xffff
LINK1_MISSED_COMMA_CNT          0xffff
LINK0_OVERFLOW_CNT              0x0
LINK1_OVERFLOW_CNT              0x0
LINK0_UNDERFLOW_CNT             0xffff
LINK1_UNDERFLOW_CNT             0xffff
LINK0_SBIT_OVERFLOW_CNT         0xffff
LINK1_SBIT_OVERFLOW_CNT         0xffff
----------OH3----------
LINK0_MISSED_COMMA_CNT          0xffff
LINK1_MISSED_COMMA_CNT          0xffff
LINK0_OVERFLOW_CNT              0x0
LINK1_OVERFLOW_CNT              0x0
LINK0_UNDERFLOW_CNT             0xffff
LINK1_UNDERFLOW_CNT             0xffff
LINK0_SBIT_OVERFLOW_CNT         0xffff
LINK1_SBIT_OVERFLOW_CNT         0xffff
----------OH6----------
LINK0_MISSED_COMMA_CNT          0xffff
LINK1_MISSED_COMMA_CNT          0xffff
LINK0_OVERFLOW_CNT              0x0
LINK1_OVERFLOW_CNT              0x0
LINK0_UNDERFLOW_CNT             0xffff
LINK1_UNDERFLOW_CNT             0xffff
LINK0_SBIT_OVERFLOW_CNT         0xffff
LINK1_SBIT_OVERFLOW_CNT         0xffff
Trigger link of OHs: [2, 3, 6] failed, but I was told to accept bad trigger links
FPGA Communication Established
====================
Step 4: Checking VFAT Communication
====================
...
...
```

And also with `checkCSCTrigLink`:

```bash
% testConnectivity.py -d --shelf=2 -s5 -o 0x10 --nPhaseScans=100 --skipDACScan --skipScurve --checkCSCTrigLink -a
...
...
====================
Step 3: Programming FPGA & Checking FPGA Communication
====================
Initial value to write: 1, register GEM_AMC.TTC.GENERATOR.ENABLE
Initial value to write: 1, register GEM_AMC.TTC.GENERATOR.SINGLE_HARD_RESET
Initial value to write: 0, register GEM_AMC.TTC.GENERATOR.ENABLE
Checking trigger link status:
--=======================================--
-> GEM SYSTEM TRIGGER LINK INFORMATION
--=======================================--

----------OH4----------
LINK0_MISSED_COMMA_CNT          0xffff
LINK1_MISSED_COMMA_CNT          0x0
LINK0_OVERFLOW_CNT              0x0
LINK1_OVERFLOW_CNT              0x0
LINK0_UNDERFLOW_CNT             0xffff
LINK1_UNDERFLOW_CNT             0x5
LINK0_SBIT_OVERFLOW_CNT         0xffff
LINK1_SBIT_OVERFLOW_CNT         0x0
----------OH5----------
LINK0_MISSED_COMMA_CNT          0x0
LINK1_MISSED_COMMA_CNT          0xffff
LINK0_OVERFLOW_CNT              0x0
LINK1_OVERFLOW_CNT              0x0
LINK0_UNDERFLOW_CNT             0x4
LINK1_UNDERFLOW_CNT             0xffff
LINK0_SBIT_OVERFLOW_CNT         0x0
LINK1_SBIT_OVERFLOW_CNT         0xffff
Trigger link of OHs: [4, 5] failed, but I was told to accept bad trigger links
FPGA Communication Established
====================
Step 4: Checking VFAT Communication
====================
...
...
```

**Comment by bdorney** [comment]
Review needed @mexanick @BenjaminRS @jsturdy

**Review by BenjaminRS — APPROVED** [review]
ok, as this has been tested and works I believe it can be approved

**Review by jsturdy — APPROVED** [review]
Cursory, ideal candidate for significant cleanup during post legacy migration

**Comment by bdorney** [comment]
> Cursory, ideal candidate for significant cleanup during post legacy migration

Not sure what that means; but sounds like a concern for another day :D

**Comment by jsturdy** [comment]
> Not sure what that means; but sounds like a concern for another day :D

It means that the looping over links could be significantly cleaned up with some proper lambda function (and the display of links as a table should be done), but yes, another day (but think about that for future contributions all, do it nice and right now so it doesn't have to be redone later!)

**Pull request merged**